sympy/vector/functions.py

```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
     scalar : SymPy Expr
         The scalar field to compute the gradient of
 
-    coord_sys : CoordSysCartesian
-        The coordinate system to calculate the gradient in
-
     Examples
     ========
 
-    >>> from sympy.vector import CoordSysCartesian, gradient
-    >>> R = CoordSysCartesian('R')
+    >>> from sympy.vector import CoordSystem3D, gradient
+    >>> R = CoordSystem3D('R')
     >>> s1 = R.x*R.y*R.z
-    >>> gradient(s1, R)
+    >>> gradient(s1)
     R.y*R.z*R.i + R.x*R.z*R.j + R.x*R.y*R.k
     >>> s2 = 5*R.x**2*R.z
-    >>> gradient(s2, R)
+    >>> gradient(s2)
     10*R.x*R.z*R.i + 5*R.x**2*R.k
 
     """
+    # Check input
+    if isinstance(scalar, Zero) or scalar == 0:
+        return Vector.zero
+
+    if not issubclass(type(scalar), Expr) or scalar.is_Vector:
+        raise TypeError("Your input should be a scalar field.")
+
+    for atom in scalar.atoms():
+        if isinstance(atom, BaseScalar):
+            system = atom.system
+            break
```

Comment:
> Use classes to correspond to mathematical objects.

Nicely put.  Well, what do we need to do to get your original PR through?

What I’m thinking:
1. Have to depreciate CoordSysCartesian().
2. Add parser for differential_class for spherical and cylindrical coordinate systems.
3. scalar_map() will need to be fixed

Example (for item 3):
In [1]: C = CoordSystem3D('C')
In [2]: spherical = CoordSystem3D('spherical', parent=C...) #Fixes origin at C's origin
...
In   [5]: spherical.scalar_map(C)
Out[5]: {spherical_phi: C_z, spherical_r: C_x, spherical_theta: C_y}

Let me know what you think and I’ll work to get the commits in.  I will fix and add the updated *functions.py* file in a new PR once yours is through.

Future:
* Need to add handling for spherical coordinates when integrating.  Not sure where this should take place just yet.
* Needs to properly handle laplacian(1/r, spherical)